plugin/trino-kudu/src/main/java/org/apache/kudu/client/KuduOperationApplier.java

```diff
@@ -13,28 +13,111 @@
+
+    private int currentOperationsInBuffer;
+    private final KuduSession kuduSession;
+
+    private KuduOperationApplier(KuduSession kuduSession)
     {
+        kuduSession.setFlushMode(MANUAL_FLUSH);
+        kuduSession.setMutationBufferSpace(bufferMaxOperations);
+        this.kuduSession = kuduSession;
+        currentOperationsInBuffer = 0;
     }
 
-    public static OperationResponse applyOperationAndVerifySucceeded(KuduSession kuduSession, Operation operation)
+    public static KuduOperationApplier fromKuduClientWrapper(KuduClientWrapper kuduClientWrapper)
+    {
+        KuduSession session = kuduClientWrapper.newSession();
+        return new KuduOperationApplier(session);
+    }
+
+    public static KuduOperationApplier fromKuduClientSession(KuduClientSession kuduClientSession)
+    {
+        KuduSession session = kuduClientSession.newSession();
+        return new KuduOperationApplier(session);
+    }
+
+    public void applyOperationAsync(Operation operation)
             throws KuduException
     {
+        if (currentOperationsInBuffer >= bufferMaxOperations) {
+            List<OperationResponse> operationResponses = kuduSession.flush();
+            verifyNoErrors(operationResponses);
+        }
         OperationResponse operationResponse = kuduSession.apply(operation);
-        if (operationResponse != null && operationResponse.hasRowError()) {
-            throw new TrinoException(GENERIC_INTERNAL_ERROR, format("Error while applying kudu operation %s: %s",
-                    operation.getChangeType().toString(), operationResponse.getRowError()));
+        checkState(operationResponse == null, "KuduSession must be configured with MANUAL_FLUSH mode");
+        currentOperationsInBuffer += 1;
+    }
+
+    private void verifyNoErrors(List<OperationResponse> operationResponses)
+    {
+        List<OperationResponseWithError> operationResponseWithErrors = operationResponses.stream()
+                .map(OperationResponseWithError::fromOperationResponse)
+                .filter(Optional::isPresent)
+                .map(Optional::get)
+                .collect(Collectors.toList());
+
+        if (!operationResponseWithErrors.isEmpty()) {
+            OperationResponseWithError firstError = operationResponseWithErrors.get(0);
+            throw new TrinoException(GENERIC_INTERNAL_ERROR, format("Error while applying %s kudu operation(s); First error: %s: %s",
+                    operationResponseWithErrors.size(),
+                    firstError.operationResponse.getOperation().getChangeType().toString(),
+                    firstError.rowError));
+        }
+        currentOperationsInBuffer = 0;
+    }
+
+    @Override
+    public void close()
+            throws KuduException
+    {
+        List<OperationResponse> operationResponses = kuduSession.close();
+        verifyNoErrors(operationResponses);
+    }
+
+    private static class OperationResponseWithError
```

Comment:
nit: Maybe call these `FailedOperation`. It's fields already convey that it's OperationResponseWithError to me.